CARDS-1867: As an admin, I can configure the patient identification process

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/CARDS-1867)

In implementing CARDS-1867, this also covers [CARDS-1771](https://phenotips.atlassian.net/browse/CARDS-1771), [CARDS-1772](https://phenotips.atlassian.net/browse/CARDS-1772), and [CARDS-1845](https://phenotips.atlassian.net/browse/CARDS-1845).

This commit adds a new administration page, Patient Identification, which changes how the Proms homepage works.
![image](https://user-images.githubusercontent.com/4656440/185635819-e0bf89a0-1738-44e4-a658-df367d27225d.png)

1. `Patients can answer surveys without a personalized link` will change the default login page, preventing people from logging in as a patient when they enter it.

2. `Patients must confirm their identity by providing their date of birth and either MRN or HCN` will control the patient login page. This can only be turned off if users are required to use a personalized link (i.e. 1. is disabled), as without a token we otherwise have no means of determining which patient a user is.

3. `The authentication token is valid after the associated event for:` alters how long authentication tokens provided by the email service are valid for. Prior to this PR, they were valid until midnight the day of the event. They are now set to 0 days after the event (the night of) by default.

*To test*:
1. While logged in as admin, you can alter Patient Identification settings from Administration/Patient Identification on the sidebar.
2. You can confirm the behaviour of 1. by altering it and logging out. The patient login will no longer be available.
3. You can confirm the behaviour of 2. by generating a patient login token. To do this, create a Patient Information form and a Visit Information form (or import some from the Proms importer using the instructions at #979 ). Then, try logging in with the token by visiting localhost:8080/Proms/?auth_token=<token>.
4. You can confirm the behaviour of 3. by forcing an email to be sent (see [these instructions](https://github.com/data-team-uhn/cards/blob/dev/Utilities/Development/EmailServer/README.md)), and checking the expiry date of the token generated. Easiest way to look at all tokens generated is by visiting http://localhost:8080/jcr:system/cards:tokens.2.json .

OrdiNeu committed:
CARDS-1867: As an admin, I can configure the patient identification process
Better ordering, title and hint for the admin section

marta- committed:
CARDS-1867: As an admin, I can configure the patient identification process
* Reorganized the admin screen
* Removed unnecessary subtitle
* Made tokenLifetime a text field instead of a checkbox
* Improved usability
* Removed unnecessary imports

OrdiNeu committed:
CARDS-1867 / CARDS-1771: Add configuration for whether patients can answer surveys without a personalized link

OrdiNeu committed:
CARDS-1845: Handle post-visit surveys

OrdiNeu committed:
CARDS-1867: Handle post-visit surveys

OrdiNeu committed:
CARDS-1845: Fix the timer on QuestionnaireSet.jsx

OrdiNeu committed:
CARDS-1867: Code cleanup

OrdiNeu committed:
CARDS-1867: Fix expiry date display

marta- commented on proms-resources/clinical-data/pom.xml:
```diff
@@ -47,6 +47,8 @@
               SLING-INF/content/Homepage/Proms/Mood.xml;path:=/Proms/Mood;overwrite:=true,
               SLING-INF/content/Homepage/Proms/ClinicMapping.xml;path:=/Proms/ClinicMapping;overwrite:=true,
+              SLING-INF/content/Homepage/Proms/PatientIdentification.xml;path:=/Proms/PatientIdentification;overwrite:=true,
+              SLING-INF/content/Homepage/Proms/WelcomeMessage.xml;path:=/Proms/WelcomeMessage;overwrite:=true,
```
This might be a rebase artifact. The WelcomeMessage config has been moved to SurveyInstructions in #1134.

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/PatientIdentification.xml:
```diff
@@ -0,0 +1,37 @@
+    </property>
+    <property>
+        <name>tokenLifetime</name>
+        <value>2</value>
```
Since we switched from hours to days, for proms tokenLifetime should be 0 to keep the current behavior.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -472,6 +474,17 @@ function QuestionnaireSet(props) {
+        response.ok ? response.json() : Promise.resolve({tokenLifetime: 0}) // At midnight the day-of by default
+      ).then((json) => {
+        // Convert that number of days from now into midnight
+        let lifetime = DateTime.utc().plus({days: json["tokenLifetime"]}).endOf('day').diff(DateTime.utc(), 'hours').hours;
```
Is this going to give an accurate results when a patient starts filling out their post-discharge (patient experience) survey at 11:05 pm with a token expiring at midnight?

marta- replied:
Addressed in c253896.

OrdiNeu committed:
CARDS-1867: Remove rebase artifacts

marta- commented on proms-resources/clinical-data/pom.xml:
```diff
@@ -47,8 +47,6 @@
               SLING-INF/content/Homepage/Proms/Cardio.xml;path:=/Proms/Cardio;overwrite:=true,
               SLING-INF/content/Homepage/Proms/Mood.xml;path:=/Proms/Mood;overwrite:=true,
               SLING-INF/content/Homepage/Proms/ClinicMapping.xml;path:=/Proms/ClinicMapping;overwrite:=true,
-              SLING-INF/content/Homepage/Proms/PatientIdentification.xml;path:=/Proms/PatientIdentification;overwrite:=true,
```
No, PatientIdentification should stay, that's the new config we introduced.

OrdiNeu committed:
CARDS-1867: Accidentally cut out an extra line

sdumitriu reviewed (commented):
Move all the configuration reading to a separate service component.

OrdiNeu committed:
CARDS-1867: Move config reading to a new Util

OrdiNeu committed:
CARDS-1771: Remove the ability for users to see the Proms homepage
withotu a personalized link

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/index.jsx:
```diff
@@ -48,6 +53,19 @@ function PromsHomepage (props) {
+      .then((response) => response.ok ? response.json() : Promise.reject(response))
+      .then((json) => {
+        let auth_token = new URLSearchParams(window.location.search).get(AUTH_TOKEN_PARAM);
+        if (!(json[ENABLED_PROP] || auth_token)) {
```
Can we rename `ENABLED_PROP` to something that makes it clear what is enabled? Otherwise I'm fine with just using `json.enableTokenlessAuth` directly since this prop name doesn't look like it will be used more than once.

OrdiNeu replied:
Done in [7246b80](https://github.com/data-team-uhn/cards/pull/1157/commits/7246b8085b60b5ad94b778622778661a0ca26f85) :heavy_check_mark:

OrdiNeu committed:
CARDS-1867: Apply code review suggestion by @marta-

marta- committed:
CARDS-1867 / CARDS-1845: Handle post-visit surveys
More accurate computation of the token expiry time

marta- committed:
CARDS-1867: As an admin, I can configure the patient identification process
* Friendlier-looking error page displayed when tokenless auth is disabled
* Includes removal of the anchor link ffrom all markdown-generated headings

marta- committed:
CARDS-1867: As an admin, I can configure the patient identification process
UX improvement: signing out as a patient now sends the patient to the patient portal landing page instead of the cards login page

marta- commented:
This PR cannot be tested as is because of CARDS-1901, which appears to be fixed by PR #1166. Putting this one _on hold_ until #1166 is merged and 1867 can be rebased for proper testing.

marta- reviewed (approved):
Looks fine to me. I did not test the token lifetime setting.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+     * @return True if tokenless authentication is enabled, {@code false} if the configuration could not be found or
+     *         it is disabled
+     */
+    boolean tokenlessAuthEnabled();
```
The java naming convention is for methods to have a verb in their name.

```suggestion
    boolean isTokenlessAuthEnabled();
```

sdumitriu replied:
```suggestion
    boolean isTokenlessAuthEnabled();
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+     * @return True if patient identification is required or if the configuration could not be found,
+     *         {@code false} otherwise
+     */
+    boolean patientIdentificationRequired();
```
```suggestion
    boolean isPatientIdentificationRequired();
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return The amount of time after an appointment ends that tokens to questionnaires should be valid for
+     */
+    int tokenLifetime();
```
```suggestion
    int getTokenLifetime();
```

But this is too technical, bound to the implementation detail that we use tokens for authenticating a patient visit. It should have a meaningful name, something like:

```suggestion
    int getAllowedPostVisitCompletionTime();
```

Not entirely happy with this suggestion, can anybody think of a better name?

sdumitriu replied:
This raises another issue: completion time is not exactly an "authentication" issue, so should the interface have a different name? Should it be `PatientEndpointConfiguration` or `PatientAccessConfiguration`? Either way, I would get rid of `Utils` at the end.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/AbstractEmailNotification.java:
```diff
@@ -47,19 +48,29 @@
+    private static final String CONFIG_NODE = "/Proms/PatientIdentification";
+
+    private static final String TOKEN_LIFETIME_PROP = "tokenLifetime";
+
```
These are not used, remove.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Obtain the amount of time after an appointment ends that tokens to questionnaires should be valid for.
+     *
+     * @return The amount of time after an appointment ends that tokens to questionnaires should be valid for
```
What is this amount measured in?

```suggestion
     * @return a number of days
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+    boolean patientIdentificationRequired();
+
+    /**
+     * Obtain the amount of time after an appointment ends that tokens to questionnaires should be valid for.
```
```suggestion
     * Get the configured amount of time, in days, after an appointment ends that the patient can still fill in the
       visit surveys. {@code -1} means the patient can fill in the forms until the midnight right before the visit,
       {@code 0} means until the midnight right after the visit, {@code 7} means until one week after the visit.
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+package io.uhndata.cards.proms.api;
+
+/**
+ * Basic utilities for grabbing configuration details from the patient identification node.
```
How the configuration is stored is an implementation detail. We may switch this to OSGI config in the future, or there may be multiple configurations for different endpoints. The important thing is that this is the patient authentication configuration.

```suggestion
 * Configuration for how the patient can access the patient-facing UI.
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -85,6 +86,16 @@ public class ValidateCredentialsServlet extends SlingAllMethodsServlet
+
+    private static final String TOKENLESS_AUTH_ENABLED_PROP = "enableTokenlessAuth";
+
+    private static final String PATIENT_IDENTIFICATION_ENABLED_PROP = "requirePIIAuth";
```
Not used here.

sashaandjic reviewed (approved):
tested

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+    Boolean PATIENT_IDENTIFICATION_REQUIRED_DEFAULT = true;
+
+    /** The number of days a token is valid for by default (used in case of errors). */
+    int TOKEN_LIFETIME_DEFAULT = 0;
```
As these are also specific to the implementation detail that a node is used, and not actually referenced outside the implementation, I would move them to the implementation class.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAuthConfigUtils.java:
```diff
@@ -0,0 +1,71 @@
+public interface PatientAuthConfigUtils
+{
+    /** The location of the configuration node for patient auth. */
+    String CONFIG_NODE = "/Proms/PatientIdentification";
```
This is a bit problematic, since we already know we at least want to change the `/Proms` name to something else, but may also add more than one endpoint, each with its own configuration. In the next version we need to come up with a way to indicate which endpoint's configuration we want to access.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -142,10 +157,11 @@ private void handleTokenlessAuthentication(final SlingHttpServletRequest request
         final SlingHttpServletResponse response, final Session session, final ResourceResolver rr)
         throws IOException, RepositoryException
     {
+        // TODO: Check to see if patient authentication is enabled in the first place
```
Isn't this addressed in the next diff?

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -156,19 +172,19 @@ private void handleTokenlessAuthentication(final SlingHttpServletRequest request
     }
 
-    private void generateAndApplyToken(final SlingHttpServletRequest request,
+    private void generateAndApplyToken(final SlingHttpServletRequest request, final Session session,
```
The passed `session` doesn't appear to be used in the method. Can the argument be removed?

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/PatientIdentification.xml:
```diff
@@ -0,0 +1,37 @@
+<node>
+    <name>PatientIdentification</name>
+    <property>
+        <name>enableTokenlessAuth</name>
```
Usually names starting with verbs are for methods. Can these be renamed in a passive voice: `tokenlessAuthEnabled`, `piiAuthRequired`.

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/PatientIdentification.xml:
```diff
@@ -0,0 +1,37 @@
+        <type>Boolean</type>
+    </property>
+    <property>
+        <name>tokenLifetime</name>
```
`allowedPostVisitCompletionTime`?

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/PatientIdentificationConfiguration.jsx:
```diff
@@ -0,0 +1,125 @@
+  const labels = {
+    enableTokenlessAuth: "Patients can answer surveys without a personalized link",
+    requirePIIAuth: "Patients must confirm their identity by providing their date of birth and either MRN or HCN",
+    tokenLifetime: "The authentication token is valid after the associated event for:"
```
We shouldn't mention tokens here either.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/PromsLandingPage.jsx:
```diff
@@ -64,10 +64,18 @@ function PromsLandingPage(props) {
-    setIsOpen(!(userType == USER_TYPE_HCP));
+
+    // If tokenless auth is disabled or if the USER_TYPE_PARAM is set, we do not display ourselves
+    fetch(CONFIG)
```
1.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -472,6 +471,16 @@ function QuestionnaireSet(props) {
 
+  function fetchVisitExpiry() {
+    // Memoize expiryOffset so we don't poll it more often than necessary
+    return fetch("/Proms/PatientIdentification.json")
```
2.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/index.jsx:
```diff
@@ -48,6 +53,19 @@ function PromsHomepage (props) {
 
+  // Fetch tokenless auth
+  useEffect(() => {
+    fetch(CONFIG)
```
3.

This is smelly. Is there a way to not have the same thing fetched 3 times in 3 different places?

marta- replied:
Fetch once in index.jsx and pass the response down to PatientIdentification and QuestionnaireSet? We do something similar for other configs.

OrdiNeu replied:
I passed it onto `QuestionnaireSet`. Unfortunately, `PromsLandingPage` and `proms/index.jsx` don't appear to share any heirarchy as far as I can see.

OrdiNeu committed:
CARDS-1867: Apply code review comments by @sdumitriu

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -93,7 +93,7 @@ const useStyles = makeStyles(theme => ({
 
 function QuestionnaireSet(props) {
-  const { subject, username, displayText, contentOffset } = props;
+  const { subject, username, displayText, contentOffset, expiryOffset } = props;
```
`contentOffset, expiryOffset` - two props with VERY different purposes and the same suffix. This line of code is going to be very confusing to read in a couple of months.

OrdiNeu committed:
CARDS-1867: Apply code review suggestion by @marta-

OrdiNeu committed:
CARDS-1867: Apply further code suggestions by @sdumitriu

OrdiNeu committed:
CARDS-1867: Codestyle

OrdiNeu committed:
CARDS-1867: Patient authentication is always required if tokenless auth
is disabled

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/api/PatientAccessConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Get the configured amount of time, in days, after an appointment ends that the patient can still fill in the
+     * visit surveys. {@code -1} means the patient can fill in the forms until the midnight right before the visit,
+     * {@code 0} means until the midnight right after the visit, {@code 7} means until one week after the visit.
```
@marta-  Do we need an "infinity" option, e.g. always allowed to fill in regardless of when the visit was?

marta- replied:
Not a bad idea for future proofing, but not urgent.

sdumitriu reviewed (commented):
Code looks good, didn't test it yet.

marta- committed:
CARDS-1867: As an admin, I can configure the patient identification process
* Ensure that there's always at least one patient identification mode set when configuring the patient identification process from the admin screen
* Minor label rewording on the configuration admin screen

sashaandjic commented:
QUESTION: Shouldn't the following setting allow patient to login with auth token and without DOB and MRN#?
![image](https://user-images.githubusercontent.com/23243244/189450999-bf1f9380-7cf3-4aea-a88a-9746c6fd3cfe.png)

marta- committed:
CARDS-1867: Patient authentication is always required if tokenless auth is disabled - fix logic bug introduced in cc157e2

marta- commented:
> QUESTION: Shouldn't the following setting allow patient to login with auth token and without DOB and MRN#? With the current implementation, I have to do both.

Fixed in c2bf84d.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
There is a small regression, that will most likely never happen in production.
To reproduce:
- generate token  for an incoming visit
- go to `localhost:8080/?auth_token=xxxxx` ( insetad of localhost:8080/**Proms.html/**?auth_token=xxxxx)
- the following screen will appear as expected
![image](https://user-images.githubusercontent.com/23243244/189475713-25e494c5-9549-4660-8c87-91c1d2710427.png)
- click on `GO TO MY SURVEYS` button
- the expected behavior would be landing on either survey's screen or patient login page (depending on the admin's setting), but instead I get this:
![image](https://user-images.githubusercontent.com/23243244/189475823-73cb80bf-a7ea-45a8-9ee1-0818f7f1f8c9.png)
